### PR TITLE
Use enum for compression algorithm

### PR DESCRIPTION
**Pull request by MinnDevelopment**
[contributing]: https://github.com/DV8FromTheWorld/JDA/wiki/5%29-Contributing

## Pull Request Etiquette

- [x] I have checked the PRs for upcoming features/bug fixes.
- [x] I have read the [contributing guidelines][contributing].

### Changes

- [x] Internal code
- [x] Library interface (affecting end-user code) 
- [ ] Documentation
- [ ] Other: \_____ 

Closes Issue: NaN

## Description

Discord might add other algorithms to choose from like [zstd](https://github.com/luben/zstd-jni) in the future.

### CODE REVIEW

File: src/main/java/net/dv8tion/jda/api/JDABuilder.java [hunk 1]
```diff
@@ -216,25 +217,26 @@ public JDABuilder setContextEnabled(boolean enable)
     }
 
     /**
-     * Enable stream-compression on the gateway connection,
+     * Enable compression on the gateway connection,
```
Comment (DV8FromTheWorld) [inline src/main/java/net/dv8tion/jda/api/JDABuilder.java]:
```suggestion
     * Set compression type used with the gateway connection,
```

File: src/main/java/net/dv8tion/jda/api/JDABuilder.java [hunk 2]
```diff
@@ -216,25 +217,26 @@ public JDABuilder setContextEnabled(boolean enable)
...
      * this will decrease the amount of used bandwidth for the running bot instance
      * for the cost of a few extra cycles for decompression.
-     * <br><b>Default: true</b>
+     * <br><b>Default: {@link net.dv8tion.jda.api.utils.Compression#ZLIB}</b>
      *
      * <p><b>We recommend to keep this enabled unless you have issues with the decompression</b>
```
Comment (DV8FromTheWorld) [inline src/main/java/net/dv8tion/jda/api/JDABuilder.java]: The verbiage on this probably needs to be updated.

File: src/main/java/net/dv8tion/jda/api/JDABuilder.java [hunk 3]
```diff
@@ -216,25 +217,26 @@ public JDABuilder setContextEnabled(boolean enable)
...
      * <br>This mode might become obligatory in a future version, do not rely on this switch to stay.
```
Comment (DV8FromTheWorld) [inline src/main/java/net/dv8tion/jda/api/JDABuilder.java]: This comment is effectively useless. Lets nix it.

File: src/main/java/net/dv8tion/jda/api/JDABuilder.java (hunk 2 above)
Comment (DV8FromTheWorld) [reply src/main/java/net/dv8tion/jda/api/JDABuilder.java]: Also explain how you would disable it now (`false` vs new `NONE` enum value)

File: src/main/java/net/dv8tion/jda/api/sharding/DefaultShardManagerBuilder.java [hunk 4]
```diff
@@ -218,25 +219,25 @@ public DefaultShardManagerBuilder setContextEnabled(boolean enable)
...
+     * @param  compression
+     *         The compression algorithm to use for the gateway connection
      *
      * @return The DefaultShardManagerBuilder instance. Useful for chaining.
      *
      * @see    <a href="https://discordapp.com/developers/docs/topics/gateway#transport-compression" target="_blank">Official Discord Documentation - Transport Compression</a>
      */
     @Nonnull
-    public DefaultShardManagerBuilder setCompressionEnabled(boolean enable)
+    public DefaultShardManagerBuilder setCompressionEnabled(Compression compression)
```
Comment (DV8FromTheWorld) [inline src/main/java/net/dv8tion/jda/api/sharding/DefaultShardManagerBuilder.java]: Missing `@Notnull` and `Check.notNull`

Comment (DV8FromTheWorld) [reply src/main/java/net/dv8tion/jda/api/sharding/DefaultShardManagerBuilder.java]: Also method name needs to be renamed to `setCompression`

File: src/main/java/net/dv8tion/jda/api/JDABuilder.java (hunk 3 above)
Comment (MinnDevelopment) [reply src/main/java/net/dv8tion/jda/api/JDABuilder.java]: I don't see how it is useless, they might still require a specific compression in the future.

**Comment by MinnDevelopment** [comment]
I have moved all the decompression logic to a `ZlibDecompressor` which allows us to easily implement other compression algorithms later on.

**Pull request merged**